cassiopeia/solver/MaxCutSolver.py

```diff
@@ -0,0 +1,169 @@
+supertree generated over phylogenetic trees for each individual character. The
+goal is to find a partition on the graph that resolves triplets on the 
+supertree, grouping together samples that share mutations and seperating
+samples that differ in mutations.
+"""
+import itertools
+import networkx as nx
+import numpy as np
+import pandas as pd
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+from cassiopeia.solver import GreedySolver
+from cassiopeia.solver import graph_utilities
+
+
+class MaxCutSolver(GreedySolver.GreedySolver):
+    def __init__(
+        self,
+        character_matrix: pd.DataFrame,
+        missing_char: str,
+        meta_data: Optional[pd.DataFrame] = None,
+        priors: Optional[Dict] = None,
+        sdimension: Optional[int] = 3,
+        iterations: Optional[int] = 50,
+        weights: Optional[Dict] = None,
+    ):
+        """The MaxCutSolver implements a top-down algorithm that recursively
+        partitions the sample set based on connectivity. At each recursive step,
+        a connectivity graph is generated for the sample set, where shared
+        mutations are coded as strong negative connections and differing mutations
+        are coded as positive connections. Then a partition is generated by finding
+        a maximum weight cut over the graph. The final partition is also improved
+        upon by a greedy hill-climbing procedure that further improves the cut.
+
+        Args:
+            character_matrix: A character matrix of observed character states for
+                all samples
+            missing_char: The character representing missing values
+            meta_data: Any meta data associated with the samples
+            priors: Prior probabilities of observing a transition from 0 to any
+                character state
+            sdimension: The number of dimensions to use for the embedding space.
+                Acts as a hyperparameter
+            iterations: The number of iterations in updating the embeddings.
+                Acts as a hyperparameter
+            weights: A set of optional weights for edges in the connectivity graph
+
+        Attributes:
+            character_matrix: The character matrix describing the samples
+            missing_char: The character representing missing values
+            meta_data: Data table storing meta data for each sample
+            priors: Prior probabilities of character state transitions
+            tree: The tree built by `self.solve()`. None if `solve` has not been
+                called yet
+            prune_cm: A character matrix with duplicate rows filtered out, removing
+                doublets from the sample set
+            sdimension: The number of dimensions to use for the embedding space
+            iterations: The number of iterations in updating the embeddings
+            weights: A set of optional weights for edges in the connectivity graph
+        """
```

Comment:
Still not in the right place. Put this _between_ the class declaration and the init function declaration. see my code in the NeighborJoiningSolver for an example.